### PR TITLE
[halide-backend] Random number generation

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130258
* __->__ #129314

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/halide.py [hunk 1]
```diff
@@ -389,19 +390,52 @@ def bitwise_right_shift(a, b):
...
+        result = f"{result} % {size}"
+        result = f"{int64(result)} + {int64(low)}"
+        return result
 
     @staticmethod
     def load_seed(name, offset):
-        raise Unsupported("rand")
+        seed = ops.load(name, offset)
+        # Halide requires 32-bit seeds
+        return f"hl.cast(hl.Int(32), {seed} ^ ({seed} >> 32))"
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:438]: Maybe seed is better to be an unsigned int.

File: torch/_inductor/codegen/halide.py [hunk 2]
```diff
@@ -389,19 +390,52 @@ def bitwise_right_shift(a, b):
...
     def randn(seed, offset):
-        raise Unsupported("rand")
+        # Box-Muller transform
+        return ops.mul(
+            ops.sqrt(
+                ops.mul(
+                    ops.constant(-2, torch.float32), ops.log(ops.rand(seed, offset))
+                )
+            ),
+            ops.cos(
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:404]: Not sure if having a randn2x that also returns the other random number computed with ops.sin can be more efficient. But like triton's rand4x , the tricky part is how to integrate into inductor so we can properly codegen then

File: torch/_inductor/codegen/halide.py [hunk 3]
```diff
@@ -389,19 +390,52 @@ def bitwise_right_shift(a, b):
...
+                ops.mul(
+                    ops.constant(2 * math.pi, torch.float32),
+                    ops.rand(
+                        ops.bitwise_xor(seed, ops.constant(0x7777, torch.int32)), offset
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:408]: Curious how you pick 0x7777 as the mask?

**Review by shunting314 — APPROVED** [review]

File: torch/_inductor/codegen/halide.py [hunk 4]
```diff
@@ -404,19 +405,52 @@ def bitwise_right_shift(a, b):
...
     @staticmethod
     def rand(seed, offset):
-        raise Unsupported("rand")
+        return f"hl.random_float({seed} ^ {offset})"
 
     @staticmethod
     def randn(seed, offset):
-        raise Unsupported("rand")
+        # Box-Muller transform
+        return ops.mul(
```
Comment (eellison) [inline torch/_inductor/codegen/halide.py:413]: wonder if better as decomposition..

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**